File: Programmers/BFS및DFS/단어변환.py
"""
단어변환
"""
def diffOne(a,b) :
    count = 0
    for i in range(len(a)) :
        if a[i] != b[i] :
            count += 1
            if count > 1 :
                return False
    if count == 1 :
        return True
    return False 

def DFS(start, target, words, depth, visited) :
    if start == target :
        global answer
        if answer == 0 or answer > depth :
            answer = depth
    elif answer == 0 or answer > depth :
        for i in range(len(words)) :
            if not visited[i] and diffOne(start, words[i]):
                visited[i] = True
                DFS(words[i], target, words, depth+1, visited)
                visited[i] = False

def solution(begin, target, words):
    global answer
    answer = 0
    if target in words :
        visited = [ False for _ in range(len(words))]
        DFS(begin, target, words, 0, visited)
    return answer

print(solution("hit", "cog", ["hit","hot", "hog", "cog"]))

File: CodingTestLanguage/Javascript/Programmers/단어변환.js
function diffOne(str1, str2){
    let diff = 0
    for(let i = 0; i< str1.length; i++){
        if(str1[i] !== str2[i]) diff++
    }
    if(diff === 1) return true
    return false
}

function DFS(begin, target, words, depth){
    if(begin === target){
        answer = Math.min(answer, depth)
    }else if(depth < answer){
        for(let i = 0; i< words.length; i++){
            if(!visited[i] && diffOne(begin, words[i])){
                visited[i] = true
                DFS(words[i], target, words, depth+1)
                visited[i] = false
            }
        }
    }
}

let answer = 0
let visited = []
function solution(begin, target, words) {
    if(!words.includes(target)) return answer

    answer = words.length;
    visited = Array(answer).fill(false)
    DFS(begin, target, words, 0)
    return answer;
}

console.log(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))

